src/ZODB/tests/ConflictResolution.py

```diff
@@ -131,6 +138,27 @@ def checkBuggyResolve2(self):
                           self._dostoreNP,
                           oid, revid=revid1, data=zodb_pickle(obj))
 
+    def checkPersistentReferenceHashability(self):
+        db = DB(self._storage)
+
+        t0 = TransactionManager()
+        c0 = db.open(t0)
+        c0.root()['z'] = PHasher([PCounter(), "foobar"])
+        t0.commit()
+
+        t1 = TransactionManager()
+        t2 = TransactionManager()
+
+        c1 = db.open(t1)
+        c2 = db.open(t2)
+
+        c1.root()['z'][1] = "foo"
+        c2.root()['z'][1] = "bar"
+
+        t1.commit()
+        t2.commit()
+
+
```

Comment:
I'm still not convinced.

I'd like to see what happens when you end up with two different PersistentReferences that hash() to the same value but are not equal.  I'd like to see a dict/set handle that without letting that pernicious ValueError escape.

Reply:
Are you sure it is even possible? The hash depends on the oid (and the database name), so two objects can have the same hash only if they have the same oid (and the same database name).
In my understanding, this is not possible that two PersistentReference can have the same hash without being equal.

Reply:
> Are you sure it is even possible?

Theoretically, yes -- the whole point of hashing is to map a large set of values to a smaller one, and therefore hash collisions are inevitable.

In practice, maybe?  I'm not sure how tuple hashes work, but I wouldn't be surprised to find `hash((oid, '', False)) == hash((oid + constant, '', True))` for some constant value.  (I'm abusing the `+` operator here to mean arithmetic addition despite oids being constant-length bytestrings.)

Reply:
A quick search on hash collision probability convinced me that it is safe to assume any hash is unique, whatever it hashes.

Reply:
> A quick search on hash collision probability convinced me that it is safe to assume any hash is unique, whatever it hashes.

That is true for cryptographic hashes (SHA-256), somewhat less true for broken old cryptographic hashes (MD5 and SHA-1), and not true for plain old hashes such as Python's hash().

For example

```
>>> hash((-1, '', False)) == hash((-2, '', False))
True
```

which is not directly applicable here (OIDs are 8-byte bytestrings, not negative integers), but should be illustrative.

Reply:
`-1` seems to be a special case for python hash.
I am not really sure what to do to convince you now :) I cannot find a testcase where two PersistentReferences could be equal without having the same hash. Do you have an idea?

Reply:
We can always say "this is vanishingly unlikely" and let the users tell us if such collisions actually exist, by filing bugs if they get that ValueError.

Reply:
(I just don't want to be responsible for making that decision, which is why I'm declining to approve or reject this PR.  Luckily I'm not the only maintainer of ZODB!  Actually not even one of the main maintainers.)

Reply:
> We can always say "this is vanishingly unlikely" and let the users tell us if such collisions actually exist, by filing bugs if they get that ValueError.

This is how I feel :) Ascendant compatibility is assured.
Who can I summon to review this?

Reply:
`git shortlog -s|sort -n|tail` tells me the people with most commits (who probably know most about ZODB) are

```
    59	Shane Hathaway
    87	Jason Madden
    88	Marius Gedminas
   158	Barry Warsaw
   166	Tres Seaver
   281	Fred Drake
   318	Guido van Rossum
   595	Tim Peters
   899	Jeremy Hylton
  1591	Jim Fulton
```

(I've no idea how I got on that list!  Fixing tests/warnings?)